File: backend/src/api/v1/aliases.py
from typing import Literal
from uuid import UUID

from fastapi import APIRouter, Query, status
from src.core.dependencies import AliasSvc, CurrentUser
from src.schemas.alias import (
    AliasApplyRequest,
    AliasApplyResult,
    AliasCreate,
    AliasResponse,
    AliasUpdate,
)
from src.schemas.pagination import CursorPage

router = APIRouter(prefix="/api/v1/aliases", tags=["aliases"])


@router.get("", response_model=CursorPage[AliasResponse])
async def list_aliases(
    _current_user: CurrentUser,
    alias_service: AliasSvc,
    limit: int = Query(50, ge=1, le=100),
    offset: int = Query(0, ge=0),
    scope: Literal["seller", "product"] | None = Query(None),
) -> CursorPage[AliasResponse]:
    """Страница алиасов (offset-пагинация) + total — алиасов может быть много."""
    items, total = await alias_service.list_page(
        _current_user,
        scope=scope,
        limit=limit,
        offset=offset,
    )
    return CursorPage[AliasResponse](
        items=[AliasResponse.model_validate(a) for a in items],
        total=total,
    )


@router.post(
    "/apply",
    response_model=AliasApplyResult,
)
async def apply_aliases(
    current_user: CurrentUser,
    alias_service: AliasSvc,
    data: AliasApplyRequest | None = None,
) -> AliasApplyResult:
    """Применить все алиасы (или одного скоупа) к существующим записям."""
    return await alias_service.apply_all(current_user.id, data.scope if data else None)


@router.post("", response_model=AliasResponse, status_code=status.HTTP_201_CREATED)
async def create_alias(
    data: AliasCreate,
    current_user: CurrentUser,
    alias_service: AliasSvc,
) -> AliasResponse:
    """Создать алиас и сразу применить его к подходящим записям."""  # noqa: RUF002
    alias = await alias_service.create(current_user, data)
    return AliasResponse.model_validate(alias)


@router.patch("/{alias_id}", response_model=AliasResponse)
async def update_alias(
    alias_id: UUID,
    data: AliasUpdate,
    current_user: CurrentUser,
    alias_service: AliasSvc,
) -> AliasResponse:
    """Обновить алиас и переприменить его к подходящим записям."""  # noqa: RUF002
    alias = await alias_service.update(current_user, alias_id, data)
    return AliasResponse.model_validate(alias)


@router.delete("/{alias_id}", status_code=status.HTTP_204_NO_CONTENT)
async def delete_alias(
    alias_id: UUID,
    _current_user: CurrentUser,
    alias_service: AliasSvc,
) -> None:
    """Удалить алиас. Уже применённые переименования не откатываются."""
    await alias_service.delete(_current_user, alias_id)

File: frontend/src/api/aliases.ts
/**
 * Aliases domain API — real endpoints, mirrors backend OpenAPI
 * (backend/src/api/v1/aliases.py). Алиасы нормализуют названия магазинов
 * (scope=seller) и товаров (scope=product): подстрока, либо regex при
 * is_regex; приоритет решает порядок применения. Создание/изменение алиаса
 * автоматически применяет его к существующим записям; POST /aliases/apply
 * переприменяет всё разом.
 */
import { api } from './client'
import type { Alias, AliasApplyResult, AliasDraft, AliasScope, AliasUpdatePatch, CursorPage } from './types'

export function fetchAliasesPage(params: {
  scope?: AliasScope
  limit?: number
  offset?: number
} = {}): Promise<CursorPage<Alias>> {
  const qs = new URLSearchParams()
  if (params.scope) qs.set('scope', params.scope)
  if (params.limit != null) qs.set('limit', String(params.limit))
  if (params.offset != null) qs.set('offset', String(params.offset))
  const query = qs.toString()
  return api.get<CursorPage<Alias>>(`/api/v1/aliases${query ? `?${query}` : ''}`)
}

export function createAlias(draft: AliasDraft): Promise<Alias> {
  return api.post<Alias>('/api/v1/aliases', draft)
}

export function updateAlias(id: string, patch: AliasUpdatePatch): Promise<Alias> {
  return api.patch<Alias>(`/api/v1/aliases/${id}`, patch)
}

export function deleteAlias(id: string): Promise<void> {
  return api.del(`/api/v1/aliases/${id}`)
}

/** Применить все алиасы (скоуп опционален; без него — оба). */
export function applyAliases(scope?: AliasScope): Promise<AliasApplyResult> {
  return api.post<AliasApplyResult>(
    '/api/v1/aliases/apply',
    scope ? { scope } : undefined,
  )
}
